### PR TITLE
VsCodeNeovim

### PR DESCRIPTION
**Pull request by Chillee**
The neovim rewrite!

For design doc/decisions/limitations/a rough idea of what's done, check here:
https://slack-files.com/T0ENVGY1E-F62R31A5V-1be4436e6c

Pretty much everything is in `neovim.ts` and `extension.ts`

If anybody has suggestions on anything about the design doc, or they have questions about any design choices, I'd be happy to listen to them.

(While I'm working on this I'll likely be less responsive to issues. Many of the current bugs would be fixed by this pr so please understand that I have less motivation to fix them).

### CODE REVIEW

File: neovim.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  private static async getPos(name: string): Promise<[number, number]> {
+    let [row, character] = ((await Vim.nv.callFunction('getpos', [name])) as Array<number>).slice(
+      1,
+      3
+    );
+    return [row - 1, character - 1];
+  }
+
+  static async getCursorPos(): Promise<[number, number]> {
```
Comment (billyvg) [inline neovim.ts]: `Vim.nv.cursor` will return the cursor position (there is no equivalent to `getpos` though

**Comment by jcjolley** [comment]
I'm so excited for this!  Give a holler if you need an alpha tester

**Comment by Chillee** [comment]
@jcjolley Yep! I think I actually fixed that bug locally. I'm not sure what's a good way to track issues for this. Maybe make another project and track issues there?

**Comment by jcjolley** [comment]
@Chillee Yeah, I wish that forked repos had an issue tracker built by default.

**Comment by jcjolley** [comment]
@Chillee turns out, if you go to the settings of the forked repo, you can enable issue tracking.
https://softwareengineering.stackexchange.com/a/179470/260792

**Comment by Chillee** [comment]
@jcjolley I'm not sure an issue tracker is the best format for this right now actually. The extension is still in a state of rapid flux, and things like that `jk` to `<Esc>` remapping have worked/not worked at several different points in this extension.

There's still several core components that change often, so an issue tracker is too rigid imo.

**Comment by jcjolley** [comment]
That's understandable.

Hit me up if you are looking for feedback and/or bugs. I'll happily submit
them when you're ready in whatever form you prefer. I'm wild excited about
this.

**Comment by johnfn** [comment]
oh man... he did it, the absolute madman!!!

**Comment by Chillee** [comment]
@jcjolley I think the best thing to do is just discussion on the slack channel. Whenever you have time, I'd love to get feedback and discuss things with you there.

**Comment by schickling** [comment]
Wow. This might me switch to Code after all.

**Comment by DxCx** [comment]
Exactly my taught @schickling =)

**Comment by Chillee** [comment]
@schickling @DxCx That's the dream! Hopefully by the time I'm done, there'd be no reason to use Vim besides performance and the benefits of being run in a terminal shell (ssh, multiplexing with tmux, etc.)

**Comment by Chillee** [comment]
I'm wondering whether there might be some good way of sharing buffers/splits between an instance of neovim in the terminal and actually in vscode.

For example, you open up gundo or NerdTree and it opens in the integrated terminal. Currently it opens in vscode itself as a file, but that's not optimal for several reasons (we don't have practically no control over splits nor freedom in split layout, we don't get the special colors that are made for those extensions, etc.)

On the other hand, perhaps that's not a perfect solution either, as it'll take over your terminal window

We also need to figure out where to place all the extra windows. quickfix list, etc.

The other problem is the technical issues behind actually creating the neovim instance and sharing commands between the 2, not to mention any possible incompatibilites with extensions and such.

Still, I'd like to find some way of supporting those things as well.

**Comment by Swiiip** [comment]
Regarding the buffer/window model, I think we could compare neovim buffers with VSCode editors, and neovim windows with VSCode groups

**Comment by Chillee** [comment]
@Swiiip Doing vim buffers = vscode editors has some not preferable consequences though. For example, that would mean that doing `:q` would close all of your windows. I don't think that's usually the intention of vscode users.

Perhaps it wouldn't be an issue though... If you turn off tabs, the model seems pretty similar to vim. However, then you wouldn't have tab support in vscode.

It seems like there's 2 options, each with their own pros and cons.
1. Try to pretend vim buffers don't exist (basically what we do now).
2. Basically pretend vim tabs don't exist

I'll try to write up the pros and cons of each approach later, if anybody's interested.

**Comment by haolly** [comment]
Any progress ?

**Comment by Chillee** [comment]
@haolly I've been working on it occasionally. The big blocker imo before I would feel comfortable releasing this "officially" is https://github.com/neovim/neovim/pull/5269. Without it, performance in any file with language services is complete garbage.

I could include my own diff library and handle it that way, but that seems like a fairly ugly hack.

**Comment by jcjolley** [comment]
>include my own diff library and handle it that way
That sounds like no fun.  Here's hoping that PR gets merged soon.

File: package.json [hunk 2]
```diff
@@ -107,12 +107,12 @@
       {
         "key": "Backspace",
         "command": "extension.vim_backspace",
-        "when": "editorTextFocus && vim.active && !inDebugRepl"
+        "when": "editorTextFocus && vim.active &&   && !inDebugRepl"
```
Comment (kalbasit) [inline package.json]: you've got a double `&&` typo.

Comment (Chillee) [reply package.json]: Ah thanks. I was wondering why backspace wasn't working, and thought I needed to do some extra work elsewhere to fix it.

**Comment by Chillee** [comment]
We will likely be moving to a different repo soon for development purposes. Eventually, VsCodeNeovim will almost definitely be recombined into VSCodeVim (probably under a flag). However, it's good to have a separate repo for issues/PRs/releasing betas.

**Comment by Chillee** [comment]
https://github.com/Chillee/VSCodeNeovim will be where most of the development happens from now on.

**Comment by jpoon** [comment]
Going to close this PR as development of the neovim rewrite is over at: https://github.com/Chillee/VSCodeNeovim and therefore all discussion of neovim rewrite should move over there.

@Chillee do you want to move it under the VSCodeVim org? makes it more easily discoverable. You should have the rights to do that if you choose.

**Comment by karlhorky** [comment]
@Chillee sounds awesome. Will any further updates about any future merge into VSCodeVim be posted to this issue? I'll subscribe!

**Comment by ohjames** [comment]
@karlhorky No work been done on it since a month before your comment.

**Pull request closed without merging**